File: phase_cli/cmd/shell.py
import sys
import os
import subprocess
from rich.console import Console
from rich.progress import Progress, SpinnerColumn
from phase_cli.utils.phase_io import Phase
from phase_cli.utils.secret_referencing import resolve_all_secrets
from phase_cli.utils.misc import get_default_shell, get_shell_command, clean_subprocess_env, parse_bool_flag

def phase_shell(env_name=None, phase_app=None, phase_app_id=None, tags=None, path: str = '/', shell_type=None, generate_leases: str = 'true', lease_ttl: int = None):
    """
    Launches an interactive shell with environment variables set to the secrets 
    fetched from Phase for the specified environment, resolving references as needed.

    Args:
        env_name (str, optional): The environment name from which secrets are fetched. Defaults to None.
        phase_app (str, optional): The name of the Phase application. Defaults to None.
        phase_app_id (str, optional): The ID of the Phase application. Defaults to None.
        tags (str, optional): Comma-separated list of tags to filter secrets. Defaults to None.
        path (str, optional): Specific path under which to fetch secrets. Defaults to '/'.
        shell_type (str, optional): Type of shell to launch (bash, zsh, fish, powershell, etc.). Defaults to None.
    """
    phase = Phase()
    console = Console()

    try:
        with Progress(
                SpinnerColumn(),
                *Progress.get_default_columns(),
                console=console,
                transient=True,
            ) as progress:
                task1 = progress.add_task("[bold green]Fetching secrets...", total=None)        

                # Fetch all secrets
                lease_flag = parse_bool_flag(generate_leases)
                all_secrets = phase.get(env_name=env_name, app_name=phase_app, app_id=phase_app_id, tag=tags, path=path, dynamic=True, lease=lease_flag, lease_ttl=lease_ttl)
                
                # Organize all secrets into a dictionary for easier lookup
                secrets_dict = {}
                for secret in all_secrets:
                    secret_env_name = secret['environment']
                    secret_path = secret['path']
                    key = secret['key']
                    if secret_env_name not in secrets_dict:
                        secrets_dict[secret_env_name] = {}
                    if secret_path not in secrets_dict[secret_env_name]:
                        secrets_dict[secret_env_name][secret_path] = {}
                    secrets_dict[secret_env_name][secret_path][key] = secret['value']

                # Resolve all secret references
                resolved_secrets_dict = {}
                for secret in all_secrets:
                    # Attempt to resolve secret references in the value
                    if secret["value"] is None:
                        continue
                    resolved_value = resolve_all_secrets(secret["value"], all_secrets, phase, secret.get('application'), secret.get('environment'))
                    resolved_secrets_dict[secret["key"]] = resolved_value

                # Count and get environment from the secrets for the message
                secret_count = len(resolved_secrets_dict)
                
                # Extract application and environment names
                applications = set(secret.get('application') for secret in all_secrets if secret['key'] in resolved_secrets_dict and secret.get('application'))
                environments = set(secret.get('environment') for secret in all_secrets if secret['key'] in resolved_secrets_dict)
                
                application_message = ', '.join(applications)
                environment_message = ', '.join(environments)

                secrets_env = clean_subprocess_env()

                # Overlay resolved secrets (cast to str and ignore Nones)
                for k, v in resolved_secrets_dict.items():
                    if v is None:
                        continue
                    secrets_env[str(k)] = str(v)

                # Set a PHASE_* environment variable for shell scripts to detect
                secrets_env['PHASE_SHELL'] = 'true'
                if environments:
                    secrets_env['PHASE_ENV'] = list(environments)[0]
                if applications:
                    secrets_env['PHASE_APP'] = list(applications)[0]

                # Ensure TERM is present for proper line-editing/rendering in shells like zsh
                if not secrets_env.get('TERM'):
                    secrets_env['TERM'] = 'xterm-256color'
                
                # Determine which shell to launch
                if shell_type:
                    shell_cmd = get_shell_command(shell_type)
                else:
                    shell_cmd = get_default_shell()

                if not shell_cmd:
                    console.log("[bold red]Error:[/] Unable to determine shell to launch. Please specify a shell using --shell.")
                    sys.exit(1)

                shell_name = os.path.basename(shell_cmd[0])

                # Stop the fetching secrets spinner
                progress.stop()

                # Print the message with the number of secrets injected
                if path and path != '/':
                    console.log(f"🐚 Initialized [bold green]{shell_name}[/] with [bold magenta]{secret_count}[/] secrets from Application: [bold cyan]{application_message}[/], Environment: [bold green]{environment_message}[/], Path: [bold yellow]{path}[/]")
                else:
                    console.log(f"🐚 Initialized [bold green]{shell_name}[/] with [bold magenta]{secret_count}[/] secrets from Application: [bold cyan]{application_message}[/], Environment: [bold green]{environment_message}[/]")
                
                console.log(f"[bold yellow]Remember:[/] Secrets are only available in this session. Type [bold]exit[/] or press [bold]Ctrl+D[/] to exit.\n")
                
                # Launch the interactive shell
                try:
                    subprocess.run(shell_cmd, env=secrets_env, shell=False)
                    console.log(f"\n[bold red]🐚 Shell session ended.[/] Phase secrets are no longer available.")
                except Exception as e:
                    console.log(f"[bold red]Error launching shell:[/] {e}")
                    sys.exit(1)

    except ValueError as e:
        console.log(f"Error: {e}")
        sys.exit(1)
    except KeyboardInterrupt:
        console.log("\n[bold yellow]Shell launch interrupted.[/]")
        sys.exit(0)

File: src/cmd/shell.go
package cmd

import (
	"fmt"
	"os"
	"os/exec"
	"strings"

	"github.com/phasehq/cli/pkg/phase"
	"github.com/phasehq/cli/pkg/util"
	sdk "github.com/phasehq/golang-sdk/phase"
	"github.com/spf13/cobra"
)

var shellCmd = &cobra.Command{
	Use:   "shell",
	Short: "🐚 Launch a sub-shell with secrets as environment variables",
	RunE:  runShell,
}

func init() {
	shellCmd.Flags().String("env", "", "Environment name")
	shellCmd.Flags().String("app", "", "Application name")
	shellCmd.Flags().String("app-id", "", "Application ID")
	shellCmd.Flags().String("tags", "", "Filter by tags")
	shellCmd.Flags().String("path", "/", "Path filter")
	shellCmd.Flags().String("shell", "", "Shell to use (bash, zsh, fish, sh, powershell, pwsh, cmd)")
	shellCmd.Flags().String("generate-leases", "true", "Generate leases for dynamic secrets")
	shellCmd.Flags().Int("lease-ttl", 0, "Lease TTL in seconds")
	rootCmd.AddCommand(shellCmd)
}

func runShell(cmd *cobra.Command, args []string) error {
	envName, _ := cmd.Flags().GetString("env")
	appName, _ := cmd.Flags().GetString("app")
	appID, _ := cmd.Flags().GetString("app-id")
	tags, _ := cmd.Flags().GetString("tags")
	path, _ := cmd.Flags().GetString("path")
	shellType, _ := cmd.Flags().GetString("shell")
	generateLeases, _ := cmd.Flags().GetString("generate-leases")
	leaseTTL, _ := cmd.Flags().GetInt("lease-ttl")

	appName, envName, appID = phase.GetConfig(appName, envName, appID)

	p, err := phase.NewPhase(true, "", "")
	if err != nil {
		return err
	}

	opts := sdk.GetOptions{
		EnvName: envName,
		AppName: appName,
		AppID:   appID,
		Tag:     tags,
		Path:    path,
		Dynamic: true,
		Lease:   util.ParseBoolFlag(generateLeases),
	}
	if cmd.Flags().Changed("lease-ttl") {
		opts.LeaseTTL = &leaseTTL
	}

	spinner := util.NewSpinner("Fetching secrets...")
	spinner.Start()
	allSecrets, err := p.Get(opts)
	spinner.Stop()
	if err != nil {
		return err
	}

	// Resolve references
	resolvedSecrets := map[string]string{}
	for _, secret := range allSecrets {
		if secret.Value == "" {
			continue
		}
		resolvedValue := sdk.ResolveAllSecrets(secret.Value, allSecrets, p, secret.Application, secret.Environment)
		resolvedSecrets[secret.Key] = resolvedValue
	}

	// Collect env/app info for display
	apps := map[string]bool{}
	envs := map[string]bool{}
	for _, s := range allSecrets {
		if _, ok := resolvedSecrets[s.Key]; ok {
			if s.Application != "" {
				apps[s.Application] = true
			}
			envs[s.Environment] = true
		}
	}
	appNames := mapKeys(apps)
	envNames := mapKeys(envs)

	// Build environment: inherit current env, add secrets and shell markers
	envSlice := os.Environ()
	for k, v := range resolvedSecrets {
		envSlice = append(envSlice, fmt.Sprintf("%s=%s", k, v))
	}
	envSlice = append(envSlice, "PHASE_SHELL=true")
	if len(envNames) > 0 {
		envSlice = append(envSlice, fmt.Sprintf("PHASE_ENV=%s", envNames[0]))
	}
	if len(appNames) > 0 {
		envSlice = append(envSlice, fmt.Sprintf("PHASE_APP=%s", appNames[0]))
	}
	if os.Getenv("TERM") == "" {
		envSlice = append(envSlice, "TERM=xterm-256color")
	}

	// Determine shell
	var shellArgs []string
	if shellType != "" {
		shellArgs, err = util.GetShellCommand(shellType)
		if err != nil {
			return err
		}
	} else {
		shellArgs = util.GetDefaultShell()
		if shellArgs == nil {
			return fmt.Errorf("no shell found")
		}
	}

	secretCount := len(resolvedSecrets)
	shellName := shellArgs[0]
	if path != "" && path != "/" {
		fmt.Fprintf(os.Stderr, "🐚 Initialized %s with %s secrets from Application: %s, Environment: %s, Path: %s\n",
			util.BoldGreenErr(shellName),
			util.BoldMagentaErr(fmt.Sprintf("%d", secretCount)),
			util.BoldCyanErr(strings.Join(appNames, ", ")),
			util.BoldGreenErr(strings.Join(envNames, ", ")),
			util.BoldYellowErr(path))
	} else {
		fmt.Fprintf(os.Stderr, "🐚 Initialized %s with %s secrets from Application: %s, Environment: %s\n",
			util.BoldGreenErr(shellName),
			util.BoldMagentaErr(fmt.Sprintf("%d", secretCount)),
			util.BoldCyanErr(strings.Join(appNames, ", ")),
			util.BoldGreenErr(strings.Join(envNames, ", ")))
	}
	fmt.Fprintf(os.Stderr, "%s Secrets are only available in this session. Type %s or press %s to exit.\n",
		util.BoldYellowErr("Remember:"),
		util.BoldErr("exit"),
		util.BoldErr("Ctrl+D"))

	// Launch shell
	c := exec.Command(shellArgs[0], shellArgs[1:]...)
	c.Env = envSlice
	c.Stdin = os.Stdin
	c.Stdout = os.Stdout
	c.Stderr = os.Stderr

	if err := c.Run(); err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			fmt.Fprintf(os.Stderr, "%s Phase secrets are no longer available.\n", util.BoldRedErr("🐚 Shell session ended."))
			os.Exit(exitErr.ExitCode())
		}
		return err
	}
	fmt.Fprintf(os.Stderr, "%s Phase secrets are no longer available.\n", util.BoldRedErr("🐚 Shell session ended."))
	return nil
}
